8295159: DSO created with -ffast-math breaks Java floating-point arithmetic

theRealAph opened the pull request:
A bug in GCC causes shared libraries linked with -ffast-math to disable denormal arithmetic. This breaks Java's floating-point semantics.

The bug is https://gcc.gnu.org/bugzilla/show_bug.cgi?id=55522

One solution is to save and restore the floating-point control word around System.loadLibrary(). This isn't perfect, because some shared library might load another shared library at runtime, but it's a lot better than what we do now. 

However, this fix is not complete. `dlopen()` is called from many places in the JDK. I guess the best thing to do is find and wrap them all. I'd like to hear people's opinions.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295159](https://bugs.openjdk.org/browse/JDK-8295159): DSO created with -ffast-math breaks Java floating-point arithmetic (**Bug** - P4)


### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to [01f6e224](https://git.openjdk.org/jdk/pull/10661/files/01f6e224e6e7926934a089f3a7cfda5d8ece3f39)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [53388f97](https://git.openjdk.org/jdk/pull/10661/files/53388f97d1ef2a3d47d88484a9623187b44d5302)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [53388f97](https://git.openjdk.org/jdk/pull/10661/files/53388f97d1ef2a3d47d88484a9623187b44d5302)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/10661/head:pull/10661` \
`$ git checkout pull/10661`

Update a local copy of the PR: \
`$ git checkout pull/10661` \
`$ git pull https://git.openjdk.org/jdk.git pull/10661/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10661`

View PR using the GUI difftool: \
`$ git pr show -t 10661`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10661.diff">https://git.openjdk.org/jdk/pull/10661.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/10661#issuecomment-1274996639)

theRealAph commented:
Also: this patch is Linux-only. I'll ask for help from build experts to make the tests GCC-only; it's not clear to me how.

dcubed-ojdk commented:
It seems strange to me that the native library part is here:

test/hotspot/jtreg/runtime/jni/TestDenormalFloat/libfast-math.c

and the two test files are here:

test/hotspot/jtreg/compiler/floatingpoint/TestDenormalDouble.java
test/hotspot/jtreg/compiler/floatingpoint/TestDenormalFloat.java

And the two tests don't have "@run main/native"... Maybe I'm missing
something about what you're trying to test here.

dholmes-ora commented on test/hotspot/jtreg/runtime/jni/TestDenormalFloat/libfast-math.c:
```diff
@@ -0,0 +1,24 @@
+ * questions.
+ */
+
+
```
A comment as to why this file exists will help future maintainers.

dholmes-ora commented:
This appears to be a 10 year old bug in gcc. Have we ever had any issues reported because of this? Inserting a workaround now seems rather late. Any FP-using native code can potentially break Java's FP semantics if it messes with the FPU control world (ref the old Borland compilers).

Shouldn't any workaround only be needed for the internals of `System.loadLibrary` as other JDK usages of `dlopen` should know what they are opening and that they are libraries that don't have this problem?

magicus commented:
> Also: this patch is Linux-only. I'll ask for help from build experts to make the tests GCC-only; it's not clear to me how.

@theRealAph We currently have a more or less 1-to-1 relationship between OS and compiler. From a portability perspective this is not ideal, but it is also hard to keep some kind of theoretical barrier when it never is tested.

Are you worried that this is a problem that occurs with gcc on other platforms as well, or that it occurs on linux with other compilers than gcc? 

We do not support gcc on macos, windows or aix, which leaves linux as the only gcc platform in the mainline (ports in separate repos will have to handle this themselves). We do support using clang on linux instead of gcc (but do not test it regularly). But my impression here is that this is more of a gcc-problem than a linux problem..?

magicus reviewed (approved):
I think this looks okay from a build perspective. I second David's opinion that the empty file should have a comment explaining that it exists to be compiled with a special flag.

theRealAph commented:
> It seems strange to me that the native library part is here:
> 
> test/hotspot/jtreg/runtime/jni/TestDenormalFloat/libfast-math.c
> 
> and the two test files are here:
> 
> test/hotspot/jtreg/compiler/floatingpoint/TestDenormalDouble.java test/hotspot/jtreg/compiler/floatingpoint/TestDenormalFloat.java

I already moved it.

> And the two tests don't have "@run main/native"... Maybe I'm missing something about what you're trying to test here.

Will fix.

theRealAph commented:
> This appears to be a 10 year old bug in gcc. Have we ever had any issues reported because of this?

That's what is so insidious about messing with the floating-point control word: all that happens is people get slightly inaccurate results, which they don't notice.

> Inserting a workaround now seems rather late. Any FP-using native code can potentially break Java's FP semantics if it messes with the FPU control word (ref the old Borland compilers).

Indeed, but there seem to be a number of libraries out there compiled with -ffast-math. This is the kind of thing people do with highly performance-critical stuff like ray tracing, etc. But we don't want it to break Java.

> Shouldn't any workaround only be needed for the internals of `System.loadLibrary` as other JDK usages of `dlopen` should know what they are opening and that they are libraries that don't have this problem?

Maybe, but sometimes we use system- or user-provided librries.

theRealAph commented:
> > Also: this patch is Linux-only. I'll ask for help from build experts to make the tests GCC-only; it's not clear to me how.
> 
> @theRealAph We currently have a more or less 1-to-1 relationship between OS and compiler. From a portability perspective this is not ideal, but it is also hard to keep some kind of theoretical barrier when it never is tested.

I see.

> Are you worried that this is a problem that occurs with gcc on other platforms as well, or that it occurs on linux with other compilers than gcc?

The former.

> We do not support gcc on macos, windows or aix, which leaves linux as the only gcc platform in the mainline (ports in separate repos will have to handle this themselves). We do support using clang on linux instead of gcc (but do not test it regularly). But my impression here is that this is more of a gcc-problem than a linux problem..?

Probably, yes. I'm happy just to fix it on Linux.

magicus commented:
So you want to provide this dlopen wrapper for other platforms as well?

theRealAph commented:
> > This appears to be a 10 year old bug in gcc. Have we ever had any issues reported because of this?
> 
> That's what is so insidious about messing with the floating-point control word: all that happens is people get slightly inaccurate results, which they don't notice.

e.g. https://github.com/gevent/gevent/pull/1820 is for Python (SciPy) and adversely affects time to converge. I have no idea if other Java users have similar problems, but it's a bug that I'd like to go away.

shipilev commented on test/hotspot/jtreg/compiler/floatingpoint/TestDenormalFloat.java:
```diff
@@ -0,0 +1,65 @@
+ * @run main/othervm compiler.floatingpoint.TestDenormalFloat
+ */
+
+// gcc -g -ffast-math -shared -Lbuild/linux-x86_64-server-release/jdk/lib -ljava -o libfast-math.so libfast-math.c
+// LD_LIBRARY_PATH=.:build/linux-x86_64-server-release/jdk/lib:build/linux-x86_64-server-release/jdk/lib/server/ java -cp /tmp compiler.floatingpoint.TestDenormalFloat
+//
+// Result:
+// Loading libfast-math.so
+// Exception in thread "main" java.lang.AssertionError: TEST FAILED: 0.0
+//      at compiler.floatingpoint.TestDenormalFloat.testFloats(TestDenormalFloat.java:47)
+//      at compiler.floatingpoint.TestDenormalFloat.main(TestDenormalFloat.java:57)
+
```
This comment is redundant, I think. There seem to be little point to run these tests separately?

shipilev commented on test/hotspot/jtreg/compiler/floatingpoint/TestDenormalFloat.java:
```diff
@@ -0,0 +1,65 @@
+
+    public static void main(String[] args) {
+        testFloats();
+        System.out.println("Loading libfast-math.so");
```
I propose we have the same logging statement in `TestDoubles`.

shipilev reviewed (commented):
I think the Java arithmetic correctness is the thing we are after. Allowing a stray library to break it seems very wrong. The alternative to this fix would be detecting the "bad" FPU flags like FTZ/DAZ, provided we can do that without messing with a whole lot of arch-specific code, and then asking users to opt-in for potentially breaking behavior.

But I believe current patch is the minimal thing we should do. Some nits:

theRealAph commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -79,37 +79,38 @@
 # include <sys/time.h>
 # include <sys/times.h>
+# include <sys/types.h>
 # include <sys/utsname.h>
```
> Do we need to shuffle^W sort includes in this patch? I presume you'd want this patch to be cleanly backportable, which means it should probably be as point-y as it can get.

OK, I can back that part out.

iwanowww commented:
Isn't it an illustration of a more general problem we have with native code where it can mess with FP environment at any time? 

We already have similar problems with MXCSR register and provide verification logic (part of `-Xcheck:jni`) to catch modifications and support conditional restoration of MXCSR register on x86_64. x86_32 validates x87 control word when `-Xcheck:jni` is enabled.

Should we do something similar here instead?

JornVernee commented:
For upcalls on non-Windows platforms, we also save MXCSR and restore it after the call, and load a set standard value for the Java code that's about to be executed. (I'm not sure why this isn't done on Windows, tbh)

Relevant code for JNI is here:
- Downcalls: https://github.com/openjdk/jdk/blob/1961e81e02e707cd0c8241aa3af6ddabf7668589/src/hotspot/cpu/x86/macroAssembler_x86.cpp#L5002
- Upcalls: https://github.com/openjdk/jdk/blob/1961e81e02e707cd0c8241aa3af6ddabf7668589/src/hotspot/cpu/x86/stubGenerator_x86_64.cpp#L262

FPCSR is only handled on non _LP64 it looks like.

I agree with Vladimir that this seems like a general problem of foreign code potentially messing with control bits (in theory, foreign code could violate its ABI in other ways as well). It seems that both major C/C++ x64 ABIs ([1], [2], [3]) treat the control bits as non-volatile, so the callee should preserve them. This is in theory a choice of a particular ABI, but I think in general we can assume foreign code does not modify the control bits. Though, we never know for sure of course, and I suppose this is where `RestoreMXCSROnJNICalls` comes in. There's no equivalent flag for FPCSR atm AFAICS, so the answer there seems to be "just don't do it".

Following that logic: from our perspective `dlopen` violates its ABI in certain cases. Preserving the status bits across calls to `dlopen` seems like a pragmatic solution. I'm not sure how important it is to have an opt-in for the current (broken) behavior...

[1]: https://learn.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-170#fpcsr
[2]: https://learn.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-170#mxcsr
[3]: https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf

mcimadamore commented:
> Isn't it an illustration of a more general problem we have with native code where it can mess with FP environment at any time?
> 
> We already have similar problems with MXCSR register and provide verification logic (part of `-Xcheck:jni`) to catch modifications and support conditional restoration of MXCSR register on x86_64. x86_32 validates x87 control word when `-Xcheck:jni` is enabled.
> 
> Should we do something similar here instead?

I tend to agree. As others have observed, a `dlopen` call (or something with same nefarious behavior) could also happen inside JNI code. But I think the interesting (and perhaps surprising) part here is that, from the perspective of the developer, no native code has executed - only a library has been loaded (via `System::loadLibrary`).

Note also that this specific problem is triggered by `dlopen` itself, because certain libraries might have some "bad" (from the perspective of JVM) initialization code. But since we're talking about JNI, JNI_OnLoad is another potential source of problem, as its native code is executed as soon as the library is loaded (and that, too, can leave the JVM in a bad state).

iwanowww commented:
> Note also that this specific problem is triggered by dlopen itself, because certain libraries might have some "bad" (from the perspective of JVM) initialization code.

... and it leads to another question whether the JVM itself is amenable to such problems. Do we need to sanitize the environment when returning from calls into the JVM?

theRealAph commented:
> Isn't it an illustration of a more general problem we have with native code where it can mess with FP environment at any time?

Yes.

> We already have similar problems with MXCSR register and provide verification logic (part of `-Xcheck:jni`) to catch modifications and support conditional restoration of MXCSR register on x86_64. x86_32 validates x87 control word when `-Xcheck:jni` is enabled.
> 
> Should we do something similar here instead?

The problem is that this bug is very insidious: the user probably doesn't know that there's anything wrong, and almost certainly has no idea that it's anything to do with JNI.

Saving and restoring the floating-point environment across dlopen() is a compromise between adding extra code at JNI callouts, which can be expensive, and shifting the burden to the user.

theRealAph commented:
> Following that logic: from our perspective `dlopen` violates its ABI in certain cases. Preserving the control bits across calls to `dlopen` seems like a pragmatic solution. I'm not sure how important it is to have an opt-in for the current (broken) behavior...

Not at all, I'd have thought.

iwanowww commented:
> The problem is that this bug is very insidious: the user probably doesn't know that there's anything wrong, and almost certainly has no idea that it's anything to do with JNI.

I'm still trying to grasp why the current problem is something different from what we experienced before.

Some platforms (x86_32 and AArch32) provide a way to restore FP environment, but it is turned off by default. Why x86_64 case is different, so it requires the logic to be turned on by default? Does the same reasoning apply to those platforms as well and we want to have `AlwaysRestoreFPU` turned on by default?    

```
$ grep -r AlwaysRestoreFPU src/hotspot/
src/hotspot//cpu/x86/templateInterpreterGenerator_x86.cpp:  if (AlwaysRestoreFPU) {
src/hotspot//cpu/x86/sharedRuntime_x86_32.cpp:  if (AlwaysRestoreFPU) {
src/hotspot//cpu/arm/sharedRuntime_arm.cpp:  if (AlwaysRestoreFPU) {
src/hotspot//cpu/arm/templateInterpreterGenerator_arm.cpp:  if (AlwaysRestoreFPU) {
src/hotspot//share/runtime/globals.hpp:  product(bool, AlwaysRestoreFPU, false,                                    \
```

iwanowww commented:
FTR I did an exercise in source code archeology and here are my findings.

The origin of `AlwaysRestoreFPU`-related code (both in x86-32 and arm-specific code) can be traced back to [JDK-6487931](https://bugs.openjdk.org/browse/JDK-6487931) and [JDK-6550813](https://bugs.openjdk.org/browse/JDK-6550813). Though both issues manifested as JVM crashes, the underlying problem was identified as FPU control word corruption by native code.

The regression test does trigger the corruption from a JNI call (using either `_FPU_SETCW` [1] or `_controlfp` [2]), but it was deliberately limited to x86-32. 

Based on that, I conclude that the problem with FP environment corruption by native code was known before, but an opt-in solution was chosen. 

(Frankly speaking, I don't know why an opt-in solution was considered sufficient. Maybe because it was erroneously believed it can only lead to a crash at runtime?)

Considering we are now aware about insidious nature of the problem (silent result corruption), I'm inclined to propose either to turn on `AlwaysRestoreFPU` by default (and provide implementation on platforms where it is missed) or, at least, catch FPU control word corruption on native->java transitions and crash the JVM advertising `-XX:+AlwaysRestoreFPU` as a solution.

[1] https://man7.org/linux/man-pages/man3/__setfpucw.3.html
[2] https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/control87-controlfp-control87-2?view=msvc-170

dholmes-ora commented:
None of this seems adequate. If you guard the dlopen this is a race as the thread doing the dlopen could get switched out leading to corruption in other threads. But we can't just simply save/restore across JNI calls either as we may save the bad env if the dlopen just occurred! The pre-existing "fixes" in this area seem similarly inadequate (and my recollection of issues was that this was an opt-in band-aid while people migrated away from using problematic native libraries, or else got them fixed).

dholmes-ora commented:
For the benefit of the mailing list my previous comment has been edited/corrected:

Correction: I was under the false impression the FP-state was shared at the process level but I've been told it is per-thread. So the scope of the problem is much narrower.

fweimer-rh commented on test/hotspot/jtreg/compiler/floatingpoint/libfast-math.c:
```diff
@@ -0,0 +1,33 @@
+ * questions.
+ */
+
+// This file is intentionally left blank.
```
This test will silently break (no longer test what's intended) once we change GCC not to link with `crtfastmath.o` for `-shared`. Maybe you should link with `crtfastmath.o` explicitly if it exists, or change the floating point control word directly in an ELF constructor.

theRealAph replied:
Perhaps so, yes.

fweimer-rh commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -1743,7 +1744,21 @@ void * os::dll_load(const char *filename, char *ebuf, int ebuflen) {
+  int rtn = fegetenv(&curr_fenv);
+  assert(rtn == 0, "fegetnv must succeed");
   void * result = ::dlopen(filename, RTLD_LAZY);
+  rtn = fesetenv(&curr_fenv);
```
`fesetenv` in glibc does not unconditionally restore the old FPU control bits, only the non-reserved bits. (The mask is 0xf3f, which seems to cover everything in use today.)  It seems unlikely that more bits are going to be defined in the future. But using `fesetenv` to essentially guard against undefined behavior after the fact is a bit awkward.

MXSCR is passed through unconditionally.

theRealAph replied:
Mmm, but `fesetenv` is the only portable thing we have (AFAIK). An alternative to using `fesetenv` everywhere would be to use `fesetenv` in generic code and override it in a per-backend way.

fweimer-rh reviewed (commented):
I wonder if something that focuses on diagnostic tools might be better here, particularly if there hasn't been any reported breakage. The `dlopen` protection is of course very incomplete because any JNI call can change the state in unexpected ways.

On the other hand, it seems unlikely that this change breaks some undefined but intended use of the FPU state because if it is changed in `dlopen`, it's not going to be propagated across threads.

iwanowww commented:
Some additional info to consider: the library modifies only MXCSR register and `-XX:+RestoreMXCSROnJNICalls` makes the test to pass. 

```
[1] $ objdump -D libfast-math.so
...
0000000000000560 <set_fast_math>:
 560:   f3 0f 1e fa             endbr64
 564:   0f ae 5c 24 fc          stmxcsr -0x4(%rsp)
 569:   81 4c 24 fc 40 80 00    orl    $0x8040,-0x4(%rsp)
 570:   00
 571:   0f ae 54 24 fc          ldmxcsr -0x4(%rsp)
 576:   c3                      ret
```

iwanowww commented:
On non-windows x86-64 the JVM saves/restores (and adjusts if needed) MXCSR inside call stub as mandated by the ABI:
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/stubGenerator_x86_64.cpp#L262-L273

x87 is not used in x86-64-specific code anymore, so x87-specific part of FP environment (FPSR and FPCS registers) is not relevant to JVM anymore.

theRealAph commented:
So here's a thought:

Reading and writing floating-point control flags can be expensive because they're serializing operations. However, we can discover whether the processor has been put into an "odd" rounding mode with just a few floating-point instructions, so we can do:

```
if (epsilon + epsilon == 0) {
      rtn = fesetenv(&default_fenv)
    }
```

which ends up as

```
	movsd  xmm1,QWORD PTR [epsilon]
	pxor   xmm4,xmm4
	addsd  xmm1,xmm1
	ucomisd xmm0,xmm4
	jnp    ...

```

We'd need a bit more fiddling to detect changes of rounding mode as well, if we wanted to do that.

That might well be chap enough that we could do it on JNI calls. Do you think it would be worth my while doing some timings?

iwanowww commented:
That sounds like a very interesting idea. 

It would be very helpful to get an understanding how much overhead it adds in JNI stub to decide whether it's worth optimizing for.

Call stub already employs an optimization to save on writing to MXCSR:
```
    Label skip_ldmx;
    __ stmxcsr(mxcsr_save);
    __ movl(rax, mxcsr_save);
    __ andl(rax, 0xFFC0); // Mask out any pending exceptions (only check control and mask bits)
    ExternalAddress mxcsr_std(StubRoutines::x86::addr_mxcsr_std());
    __ cmp32(rax, mxcsr_std, rscratch1);
    __ jcc(Assembler::equal, skip_ldmx);
    __ ldmxcsr(mxcsr_std, rscratch1);
    __ bind(skip_ldmx);
```
According to [uops.info](https://uops.info/html-instr/STMXCSR_M32.html), latencies for `STMXCSR` vary from 7-12 cycles on Intel to up to 20 on AMD. I haven't found any details about the actual implementations in silicon (can't confirm it serializes the execution), so I'm curious how much branch prediction can hide the latency in this particular case.

If it turns out to be worth optimizing `STMXCSR` away, I see other problematic cases:

```
StubRoutines::x86::_mxcsr_std = 0x1F80;

// MXCSR.b  10987654321098765432109876543210
// 0xFFC0   00000000000000001111111111000000 // mask
// 0x1F80   00000000000000000001111110000000 // MXCSR value used by JVM
// 0x8040   00000000000000001000000001000000 // the bits -ffast-math mode unconditionally sets

// MXCSR bits:
// 15    FTZ Flush to Zero                       0 = 1
// 14:13 RC  Rounding Control                   00
// 12    PM  Precision Exception Mask            1
// 11    UM  Underflow Exception Mask            1
// 10    OM  Overflow Exception Mask             1
// 9     ZM  Zero-Divide Exception Mask          1
// 8     DM  Denormalized-Operand Exception Mask 1
// 7     IM  Invalid-Operation Exception Mask    1
// 6     DAZ Denormals Are Zeros                 0 = 1
```

The GCC bugs with `-ffast-math` only corrupts `FTZ` and `DAZ`. 

But `RC` and exception masks may be corrupted as well the same way and I believe the consequences are be similar (silent divergence in results during FP computations).

theRealAph commented:
I now have some performance results. `java.lang.foreign.CallOverheadConstant` is the test that I used to measure JNI overhead.

At present, without `-XX:+RestoreMXCSROnJNICalls`, it looks like this:

```
Benchmark                          Mode  Cnt  Score   Error  Units
CallOverheadConstant.jni_blank     avgt   40  9.968 ? 0.037  ns/op
CallOverheadConstant.panama_blank  avgt   40  8.745 ? 0.012  ns/op

```
Enabling `-XX:+RestoreMXCSROnJNICalls` makes the overhead much worse:

```
Benchmark                          Mode  Cnt   Score   Error  Units
CallOverheadConstant.jni_blank     avgt   40  14.741 ? 0.031  ns/op
CallOverheadConstant.panama_blank  avgt   40  14.620 ? 0.022  ns/op

```
and with JMH perfasm we can see why:

```
                0x00007f9f43d5698d:   sub    rsp,0x8
   1.56%        0x00007f9f43d56991:   vstmxcsr DWORD PTR [rsp]
  25.01%        0x00007f9f43d56996:   mov    eax,DWORD PTR [rsp]
  11.09%        0x00007f9f43d56999:   and    eax,0xffc0
                0x00007f9f43d5699e:   cmp    eax,DWORD PTR [rip+0xe02d234]        # 0x00007f9f51d83bd8

```
That adds 50% to the total JNI overhead. 70% to the Panama overhead.
25% of the total elapsed time is MXCSR! Reading MXCSR is expensive. So we don't do that.

So, after a lot of head scratching, I've invented an instruction sequence which doesn't read MXCSR but does a little arithmetic, and `-XX:+RestoreMXCSROnJNICalls` is:

```
CallOverheadConstant.jni_blank     avgt   40  10.675 ± 0.100  ns/op
CallOverheadConstant.panama_blank  avgt   40  10.284 ± 0.018  ns/op

```

Which is 7% added overhead for JNI, 17% for Panama. 1ns is 3.5 machine cycles: that's a bit less than the latency of a load from L1 cache.

I'm wondering if I could get away with fixing `RestoreMXCSROnJNICalls` and turning it on by default.

theRealAph commented:
> Enabling `-XX:+RestoreMXCSROnJNICalls` makes the overhead much worse:
> 
> ```
> Benchmark                          Mode  Cnt   Score   Error  Units
> CallOverheadConstant.jni_blank     avgt   40  14.741 ± 0.031  ns/op
> CallOverheadConstant.panama_blank  avgt   40  14.620 ± 0.022  ns/op
> ```

This is Zen+, by the way: Latency of stmxcsr ≤18. Better than some x86_64, worse than others.

theRealAph commented:
> Sorry, I feel like this has gone a bit off track. It started as some hardening for `loadLibrary`, but now it's about making all JNI calls a bit slower? Is there any data to suggest that this is necessary?
> 
> Would it be possible to capture some FPU state evidence in crash dumps, as an alternative?

Really? The problem is that when certain libraries are loaded, we get silently corrupted results. Vladimir Ivanov pointed out that the weakness applies to any JNI call, and we wondered if it might be possible to make the workaround so cheap that we could leave it on by default. IMVHO we probably could: the additional overhead is about 1-1.5ns. The only way to measure it is carefully written tests against a JNI call that does nothing.

theRealAph commented:
> I wonder if something that focuses on diagnostic tools might be better here, particularly if there hasn't been any reported breakage. The `dlopen` protection is of course very incomplete because any JNI call can change the state in unexpected ways.
> 
> On the other hand, it seems unlikely that this change breaks some undefined but intended use of the FPU state because if it is changed in `dlopen`, it's not going to be propagated across threads.

Why is that unlikely? `System.loadLibrary` runs on a Java thread.

iwanowww commented:
Thanks a lot for the data, Andrew.

I agree with you that the overhead `stmxcsr` introduces makes it almost prohibitive to be unconditionally placed on the hot path on the way back from a JNI call. In contrast, your idea to detect the corruption by checking the result of a carefully chosen FP expression has very modest impact while still being able to catch important types of MXCSR corruption.  I fully support having it turned on by default for JNI calls.

theRealAph commented:
On 10/26/22 03:24, John Rose wrote:
> Now I'm curious:  What's this magic code?  Does it multiply a couple of well-chosen constants and test for zero?

static const double unity = 0x1.0p-1020;
static const volatile double thresh = 0x0.0000000000003p-1022;

  if (unity + (thresh) == unity || -unity - (thresh) == -unity) {
    ... fix the badness

Here, thresh is the smallest denormal number that has two bits set. Unity
is a number such that, when thresh is added it it, must be rounded
according to the mode. These two tests detect the rounding mode in use.
If denormals are turned off (i.e. denormals-are-zero) it looks like round-
to-zero mode is in use.

This is essentially the same test that Joe Darcy posted earlier in this
thread, but scaled to make thresh so small that it is denormal.

> I said "I guess" because I'm not clear on (a) the benefit of adding that nanosecond (other than preserving denorms against a rare system fault), nor on (b) what are the remaining faults not checked for, but which a more expensive MSR spill/restore would fix.

If someone turned on flush-denormal-results-to-zero but didn't also turn
on denormals-are-zero I guess we'd miss that. Otherwise it doesn't matter.

I'd like the JVM to be protected against the GCC -ffast-math bug because the
resulting JVM fault silently produces incorrect results. We can do that with
a test after each dlopen(). But that doesn't work in a robust way: if we call
some native code that itself dlopen()s a library compiled with -ffast-math we
still lose.
Hmm, I guess I could scale the operands a bit more to detect that too... What fun!

iwanowww commented:
> At most an expanded -Xcheck:jni check for FP-state manipulation, with an
enhanced/fixed RestoreMXCSROnJNICalls is in order IMVHO.

FTR both `-Xcheck:jni` and `-XX:+RestoreMXCSROnJNICalls` already catch the problematic case being discussed.

```
$ .../jdk/bin/java ... compiler/floatingpoint/TestDenormalDouble
Exception in thread "main" java.lang.AssertionError: TEST FAILED: 0.0
        at compiler.floatingpoint.TestDenormalDouble.testDoubles(TestDenormalDouble.java:42)
        at compiler.floatingpoint.TestDenormalDouble.main(TestDenormalDouble.java:52)


$ .../jdk/bin/java ... -Xcheck:jni compiler/floatingpoint/TestDenormalDouble
Loading libfast-math.so
Java HotSpot(TM) 64-Bit Server VM warning: MXCSR changed by native JNI code, use -XX:+RestoreMXCSROnJNICall
Test passed.


$ .../jdk/bin/java ... -XX:+RestoreMXCSROnJNICalls compiler/floatingpoint/TestDenormalDouble
Loading libfast-math.so
Test passed.
```

iwanowww commented:
There's another option to consider: check MXCSR register consistency at safepoints. It's already too late to fix the corruption (the damage could be already done and the only way to proceed is to initiate a JVM crash), but it should pretty reliably catch the corruption in a prompt manner (so far, by default JVM checks and adjusts MXCSR on x86-64 only during upcalls into Java). 

Also, looks like it should mix well with the fast approximate check on JNI calls (which should be able to catch & heal some of the possible corruptions).

JornVernee commented:
I agree with David. Unconditionally doing a check on every call seems to be overkill, since it's a mostly theoretical problem at this point, and in general I think we should be able to assume that foreign code respects the ABI.

There are other things that can go wrong as well, such as foreign code installing a signal handler, which can break implicit null checks. Other things like the foreign code returning with corrupted register state, which than leads to further corruption, is also a possibility. i.e. there seem to be many more things that can go wrong if we expect native code to violate the ABI.

Even though the check can be pretty fast, we've seen that people watch the performance in this area closely, and care about every nanosecond spent here. On my own box, the `panama_blank` benchmark takes just 3.4ns, so the relative overhead could be larger depending on the machine, it seems. There was also recently a flag added to speed up native calls, namely `-XX:+UseSystemMemoryBarrier`. This could further make the relative overhead of a check larger.

All in all, I think `-Xcheck:jni` is a better place to test this kind of stuff, and encourage people to run tests with `-Xcheck:jni` before deploying to production.

But, at the same time, loading libraries is a known problematic situation, and there the performance matters far less. I'd say always checking and restoring the FPU control state, and perhaps emitting a warning message to spur people on to fix the issue in the long term, seems like a good solution to me.

theRealAph commented:
Thanks everyone for your contributions.

I think we have a good solution:
1. Add my "fast" alternative RestoreMXCSR code, and use =1 and =2 levels of checking, as suggested by John Rose. Don't enable RestoreMXCSR by default: it should be 0 (no check).
2. Warn and restore the MXCSR at safepoints. This could be a hard error rather than a warning; not sure. At least we wouldn't have silent corruption.
Everyone happy?

theRealAph commented:
That's a very interesting point. I know it used to be the case that denormals
were handled by trapping to microcode, but there are good hardware algorithms
since Schwarz et al, 2003 [1]. This paper showed how with a little hardware,
such numbers can be handled close to the speed of normalized numbers.

I deliberately ran my tests on a ten-year-old CPU, but I guess I'd have to go
further back to find a bad case.

Anyway, I plan to

a. Restore the FPU CR after calls to dlopen(3).
b. Detect FPU CR corruption at safepoints, and print a warning. At least
    the user might find out that something is wrong.

I think this will avoid most cases of badness. I guess I'll need a CSR for
this?


[1] Hardware implementations of denormalized numbers, DOI:10.1109/ARITH.2003.1207662
Conference: Computer Arithmetic, 2003. Proceedings. 16th IEEE Symposium on

JornVernee commented:
> Anyway, I plan to
> 
> a. Restore the FPU CR after calls to dlopen(3).
> b. Detect FPU CR corruption at safepoints, and print a warning. At least
> the user might find out that something is wrong.

Doing (a) seems good. I can't say for sure whether (b) is a good idea. I guess you just have some call to verify the FPU, e.g. in `ParallelCleanupTask`? I assume you don't mean to change the code for polling safepoints.

CSR seems like a good idea, since it could be a change in observable behavior. (nice to leave a paper trail I think)

theRealAph commented:
> > Anyway, I plan to
> > a. Restore the FPU CR after calls to dlopen(3).
> > b. Detect FPU CR corruption at safepoints, and print a warning. At least
> > the user might find out that something is wrong.
> 
> Doing (a) seems good. I can't say for sure whether (b) is a good idea. I guess you just have some call to verify the FPU, e.g. in `ParallelCleanupTask`? I assume you don't mean to change the code for polling safepoints.

My intention is only to do the FPU CR  check when there is an expensive operation such as moving to s safepoint. That we we won't see a significant slowdown. But we are pretty much guaranteed that the user won't get incorrect results without some sort of diagnostic.

> CSR seems like a good idea, since it could be a change in observable behavior. (nice to leave a paper trail I think)

Yep.

theRealAph commented:
I ain't dead yet.

theRealAph commented:
Reopening.

theRealAph commented:
/open

theRealAph commented:
/open

magicus reviewed (approved):
This looks good from a build perspective.

dholmes-ora commented:
Nothing seems to have changed since this was updated in October 2022 and then went through several close/open cycles. What is actually being proposed now?

theRealAph commented on test/hotspot/jtreg/compiler/floatingpoint/TestDenormalFloat.java:
```diff
@@ -0,0 +1,65 @@
+ * @test
+ * @bug 8295159
+ * @summary DSO created with -ffast-math breaks Java floating-point arithmetic
+ * @run main/othervm compiler.floatingpoint.TestDenormalFloat
```
Done.

magicus commented on make/autoconf/flags-cflags.m4:
```diff
@@ -574,7 +574,7 @@ AC_DEFUN([FLAGS_SETUP_CFLAGS_HELPER],
   # CXXFLAGS C++ language level for all of JDK, including Hotspot.
   if test "x$TOOLCHAIN_TYPE" = xgcc || test "x$TOOLCHAIN_TYPE" = xclang || test "x$TOOLCHAIN_TYPE" = xxlc; then
-    LANGSTD_CXXFLAGS="-std=c++14"
+    LANGSTD_CXXFLAGS="-std=gnu++14"
```
Uhhh, is this really okay? I thought the idea was that we should standardize on official C++, not use vendor-specific extensions. 

@kimbarrett Opinions on this?

theRealAph replied:
I pulled it out. The problem is that I have to do this:

```
    jdouble_cast(0x0030000000000000); // 0x1.0p-1020;
```

which is yucky, but not the End Of The World.

theRealAph commented:
This is a stripped-down, less risky, and simplified patch, in order to get something done.

I removed x32 handling because it have turned out to be disproportionately fiddly. The `mxcsr register we need to access is not a part of the `fenv` structure, so it needs special-case handling. This handling turns out to be problematic because HotSpot on GCC is compiled with low x32 CPU version that doesn't support the builtins or assembly instructions that we need to access `mxcsr`. I don't want to change HotSpot's default x32 CPU version, because that will affect code generation in unknown ways. Given that x32 is obsolescent, I dropped it.

In addition, I dropped the proposed warning at safepoints if the handling of denormal floating-point changes. It's fiddly and has potential compatibility issues, and may have performance impact. It's not needed to fix 8295159.

theRealAph commented:
> I'm seeing one automated test failure on Linux x86, which I don't understand because I've excluded that test for generic-i586. If anyone understands this, please shout up.

For avoidance of doubt, the test doesn't run locally, only on the Github-triggered CI.

tstuefe commented:
> > I'm seeing one automated test failure on Linux x86, which I don't understand because I've excluded that test for generic-i586. If anyone understands this, please shout up.
> 
> For avoidance of doubt, the test doesn't run locally, only on the Github-triggered CI.

Only TestDenormalFloat is excluded. The double variant still runs and fails.

tstuefe commented on src/hotspot/os/bsd/os_bsd.cpp:
```diff
@@ -972,6 +973,45 @@ bool os::dll_address_to_library_name(address addr, char* buf,
 // in case of error it checks if .dll/.so was built for the
 // same architecture as Hotspot is running on
 
+void *os::Bsd::dlopen_helper(const char *filename, int mode) {
```
I thought BSD is switching to clang.

theRealAph replied:
What difference does it make if it does?

tstuefe replied:
I was trying to understand the BSD+gcc combination. We use clang on MacOS, so the only platform I thought would be affected would be one of the BSDs. 

But I thought those also moved to clang in their builds. Hence my question.

theRealAph replied:
These third-party libraries can come from anywhere. Given that the overhead added to `dlopen()` is almost immeasurably small, I think it's better to keep it.

theRealAph commented on src/hotspot/os/bsd/os_bsd.cpp:
```diff
@@ -972,6 +973,45 @@ bool os::dll_address_to_library_name(address addr, char* buf,
+      = jdouble_cast(0x0030000000000000); // 0x1.0p-1020;
+    static const volatile double thresh
+      = jdouble_cast(0x0000000000000003); // 0x0.0000000000003p-1022;
+    if (unity + thresh == unity || -unity - thresh == -unity) {
```
`thresh` is volatile. What more is needed?

tstuefe replied:
Argh, sorry, missed that volatile.

iwanowww replied:
And I was confused at first why there's a volatile on `tresh`. A short comment describing the intentions would definitely help here.

theRealAph replied:
> And I was confused at first why there's a volatile on `tresh`. A short comment describing the intentions would definitely help here.

OK.

iwanowww commented on test/hotspot/jtreg/compiler/floatingpoint/libfast-math.c:
```diff
@@ -0,0 +1,56 @@
+#define _FPU_SETCW(fpcr) \
+  {  __asm__ __volatile__ ("msr fpcr, %0" : : "r" (fpcr)); }
+  /* Flush to zero, round to nearest, IEEE exceptions disabled.  */
+  _FPU_SETCW (_FPU_FPCR_FZ);
```
Macros make it a bit harder to decipher what happens there. At least, I'd suggest to change formatting around `_FPU_SETCW`. At first glance, it looks like it is part of the macro definition.

theRealAph replied:
OK, I see.

iwanowww reviewed (commented):
Overall, looks good. 

Primarily, I'm curious why did you decide to fix JNI on x86-64, but not on AArch64? FTR the fix covers shared library case on both architectures.

> In addition, I dropped the proposed warning at safepoints if the handling of denormal floating-point changes.
It still makes sense to introduce an assert checking that FP operations behave according to the spec. The invariant should hold for any JavaThread on any thread state transition. But I'm fine with handling it separately. 

Some minor comments follow.

magicus reviewed (approved):
The latest build changes look trivially good.

(If you decide to make other changes to the build system, please await a re-review of build changes before pushing)

theRealAph commented on test/hotspot/jtreg/compiler/floatingpoint/libfast-math.c:
```diff
@@ -0,0 +1,56 @@
+#if defined(__GNUC__)
+static void __attribute__((constructor)) set_flush_to_zero(void) {
+
+#if defined(__x86_64__) && defined(SSE)
```
I can't remember. I'll take it out.

theRealAph commented on src/hotspot/os/bsd/os_bsd.cpp:
```diff
@@ -972,6 +973,45 @@ bool os::dll_address_to_library_name(address addr, char* buf,
+      = jdouble_cast(0x0030000000000000); // 0x1.0p-1020;
+    static const volatile double thresh
+      = jdouble_cast(0x0000000000000003); // 0x0.0000000000003p-1022;
+    if (unity + thresh == unity || -unity - thresh == -unity) {
```
OK.

theRealAph commented on src/hotspot/os/bsd/os_bsd.cpp:
```diff
@@ -972,6 +973,45 @@ bool os::dll_address_to_library_name(address addr, char* buf,
 // same architecture as Hotspot is running on
 
+void *os::Bsd::dlopen_helper(const char *filename, int mode) {
+#if defined(__GNUC__)
```
Good point. I think the reason it's GCC-conditional is that a previous version used GCC extensions. I'll take it out.

theRealAph commented on test/hotspot/jtreg/compiler/floatingpoint/libfast-math.c:
```diff
@@ -0,0 +1,56 @@
+// This breaks Java's floating point arithmetic.
+
+#if defined(__GNUC__)
+static void __attribute__((constructor)) set_flush_to_zero(void) {
```
OK.

theRealAph commented on test/hotspot/jtreg/compiler/floatingpoint/libfast-math.c:
```diff
@@ -0,0 +1,56 @@
+
+#include <assert.h>
+#include <fenv.h>
+#include "jni.h"
```
Fixed.

dholmes-ora commented on src/hotspot/share/runtime/stubRoutines.cpp:
```diff
@@ -307,6 +312,28 @@ void compiler_stubs_init(bool in_compiler_thread) {
+
+  // _small_denormal is the smallest denormal number that has two bits
+  // set. _large_denormal is a number such that, when _small_denormal
+  // is added it it, must be rounded according to the mode. These two
```
s/it it, must/to it, it must/

dholmes-ora commented on test/hotspot/jtreg/compiler/floatingpoint/TestDenormalDouble.java:
```diff
@@ -0,0 +1,55 @@
+        lastDouble = 0x1.0p-1074;
+        for (double x = lastDouble * 2; x <= 0x1.0p1022; x *= 2) {
+            if (x != x || x <= lastDouble) {
+                throw new AssertionError("TEST FAILED: " + x);
```
Tests don't normally use AssertionError like this, just a plain Error or RuntimeException.

dholmes-ora commented on src/hotspot/cpu/x86/macroAssembler_x86.cpp:
```diff
@@ -5162,7 +5162,24 @@ void MacroAssembler::restore_cpu_control_state_after_jni(Register rscratch) {
     if (RestoreMXCSROnJNICalls) {
+      Label FAIL, DONE;
+#ifdef _LP64
+      // Perform a little arithmetic to make sure that denormal
+      // numbers are handled correctly, i.e. that the "Denormals Are
+      // Zeros" flag has not been set.
```
I don't understand what this part is doing. I thought it was simply checking so you could log/warn if the unexpected mode was detected. But it seems to cause MXCSR to not be restored when there is an issue, where I would expect you would always want to restore to overwrite the invalid mode the JNI call made. ??

theRealAph replied:
If we reach `FAIL`, MXCSR is reloaded from `addr_mxcsr_std()`, restoring correct IEEE behaviour. That's what RestoreMXCSROnJNICalls is supposed to do, as far as I can tell.

But I will take this part out, because there are other flags in MXCSR, so there is a potential compatibility problem if e.g. the Precision Mask were set in a JNI call.

dholmes-ora replied:
Misread the code flow. So you were skipping the restore if you didn't think it was needed? The fact the restore was now conditional is what threw me.

dholmes-ora reviewed (commented):
Meta-question and apologies if this was covered before, but why is this logic being added to stubRoutines.cpp?

theRealAph commented:
> Meta-question and apologies if this was covered before, but why is this logic being added to stubRoutines.cpp?

Because thats where @iwanowww asked me to put it. I don't much care.

iwanowww commented:
>> Meta-question and apologies if this was covered before, but why is this logic being added to stubRoutines.cpp?

> Because tha'ts where @iwanowww asked me to put it. I don't much care.

The constants were duplicated in multiple places (in particular, in `macroAssembler_x86.cpp`), so I suggested to put them in some shared place and couldn't come up with a better one than `stubRoutines.cpp/hpp`. 

But I don't see any changes in `macroAssembler_x86.cpp` anymore. What happened to them, @theRealAph?

jddarcy commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -1827,6 +1848,18 @@ void * os::Linux::dlopen_helper(const char *filename, char *ebuf,
 #endif
+
+#ifndef IA32
+    // Quickly test to make sure denormals are correctly handled.
```
Nit: I recommend using "subnormal" rather than "denormal" for general terminology on this point. While "denormal" was used in the original IEEE 754 standard from 1985, subsequent iterations of the standard using "subnormal" The term "subnormal" has also been used for the last several editions of JLS and JVMS.

theRealAph replied:
I've long avoided "subnormal" because

subnormal in British English, adjective
3.  [old-fashioned, offensive] a person of low intelligence

theRealAph replied:
All done. Good to go?

jddarcy commented on src/hotspot/share/runtime/stubRoutines.cpp:
```diff
@@ -307,6 +312,28 @@ void compiler_stubs_init(bool in_compiler_thread) {
+  // We need the addition of _large_denormal and _small_denormal to be
+  // performed at runtime. Making _small_denormal volatile ensures
+  // that the following expression isn't evaluated at compile time:
+  return (_large_denormal + _small_denormal == _large_denormal
```
As a possible future expansion, if there are cases where foreign or native code sets the rounding mode to something other than round to nearest, expressions in the same vein can be used to detect that case and restore that other aspect of the control word.

theRealAph commented:
> But I don't see any changes in `macroAssembler_x86.cpp` anymore. What happened to them, @theRealAph?

I took them out because of a potential backwards-compatibility breakage. They weren't critical to the core purpose of this patch.

iwanowww commented:
> I took them out because of a potential backwards-compatibility breakage.

Ok, I checked the removed code (https://github.com/openjdk/jdk/pull/10661/commits/b817d4757c78594be5960ee0be27013e2588d30a) and agree it is not needed here. 

`RestoreMXCSROnJNICalls` unconditionally restores MXCSR contents while the aforementioned code made it conditional based on FTZ bit fast check.  It could be considered as an optimization (to speed up `-XX:+RestoreMXCSROnJNICalls` mode), but then the fast path check should be extended to cover all possible failure modes (e.g., rounding issues).

theRealAph commented:
> > Meta-question and apologies if this was covered before, but why is this logic being added to stubRoutines.cpp?
> 
> Because tha'ts where @iwanowww asked me to put it. I don't much care.

Hi @iwanowww , do you have any suggestion about where to move the remaining code that's in stubRoutines.cpp? @dholmes-ora ?

tstuefe reviewed (approved):
This looks good to me.

One suggestion: to reduce code duplication and to make the code a bit safer against accidental returns prior to fesetenv, I would have used a mark object like this:

```
struct FenvResetMark {
  fenv_t _fenv;
  FenvResetMark()   { fegetenv(&_fenv); }
  ~FenvResetMark()  { fesetenv(&_fenv); }
};
```

and would have placed it above dlopen calls. Up to you.

---

About the dlopen calls in the JDK, at SAP we were faced with similar problems for other libc APIs (how to apply a fix to all of them). Some of these issues we solved by redirecting all calls to libjvm. Others we solved manually, in-place, with a lot of duplication. None of these sound appealing, but I like the redirect-to-libjvm route somewhat, if Oracle can be convinced.

A third option would be to use an interposition library with LD_PRELOAD. One that overwrites dlopen and redirects to the real one. I don't see this to be a practical solution but it may be valid for testing.

tstuefe commented:
One more thought, it would be good to add the FTZ_mode_enabled check to `os::run_periodic_checks()`. 

We already do signal handler checks there, and it is the right place to check for "global things third party native code may mess up". It runs when one uses `-XX:CheckJNICalls`.

iwanowww commented:
> do you have any suggestion about where to move the remaining code that's in stubRoutines.cpp? 

As it is now, `globalDefinitions.hpp/cpp` looks the most appropriate one to me. You could create new header file for it (under `share/utilities/`), but introducing one just for a single function may be an overkill.

iwanowww commented on make/test/JtregNativeHotspot.gmk:
```diff
@@ -851,6 +851,7 @@ ifeq ($(call isTargetOs, linux), true)
   BUILD_TEST_exeinvoke_exeinvoke.c_OPTIMIZATION := NONE
   BUILD_HOTSPOT_JTREG_EXECUTABLES_LIBS_exeFPRegs := -ldl
   BUILD_HOTSPOT_JTREG_LIBRARIES_LIBS_libAsyncGetCallTraceTest := -ldl
+  BUILD_HOTSPOT_JTREG_LIBRARIES_LDFLAGS_libfast-math := -ffast-math
```
Is the flag redundant by now? The test explicitly works with corresponding platform-specific registers.

theRealAph replied:
it's belt-and braces (or belt-and-suspenders in American). This way, we can still do something useful for platforms not fully supported by mainline. I don't want to add ifdefs for every platform to the test.

dholmes-ora commented:
> > do you have any suggestion about where to move the remaining code that's in stubRoutines.cpp?
> 
> As it is now, `globalDefinitions.hpp/cpp` looks the most appropriate one to me.

That seems okay to me too and much better than `stubRoutines.cpp`. Thanks.

theRealAph commented:
> One more thought, it would be good to add the FTZ_mode_enabled check to `os::run_periodic_checks()`.
> 
> We already do signal handler checks there, and it is the right place to check for "global things third party native code may mess up". It runs when one uses `-XX:CheckJNICalls`. If a native library messes with fenv, one will get a delayed assertion, with a hs-err file that lists all the shared objects.

That's a terrific idea, but maybe it'd want a CSR?

tstuefe commented:
> > One more thought, it would be good to add the FTZ_mode_enabled check to `os::run_periodic_checks()`.
> > We already do signal handler checks there, and it is the right place to check for "global things third party native code may mess up". It runs when one uses `-XX:CheckJNICalls`. If a native library messes with fenv, one will get a delayed assertion, with a hs-err file that lists all the shared objects.
> 
> That's a terrific idea, but maybe it'd want a CSR?

Maybe. CheckJNICalls is a product switch and its behavior has not changed since initial check-in AFAICS.

Since CSR-freeze is close for JDK 22, maybe this is better as a separate RFE. If you want, I can take care of it.

theRealAph commented:
> This looks good to me.
> 
> One suggestion: to reduce code duplication and to make the code a bit safer against accidental returns prior to fesetenv, I would have used a mark object like this:

Thanks. I take your point, but I think that might be something of a premature generalization right now. It would further complicate the code.

> About the dlopen calls in the JDK, at SAP we were faced with similar problems for other libc APIs (how to apply a fix to all of them). Some of these issues we solved by redirecting all calls to libjvm. Others we solved manually, in-place, with a lot of duplication. None of these sound appealing, but I like the redirect-to-libjvm route somewhat, if Oracle can be convinced.
> 
> A third option would be to use an interposition library with LD_PRELOAD. One that overwrites dlopen and redirects to the real one. I don't see this to be a practical solution but it may be valid for testing.

Thanks, interesting stuff.

theRealAph commented:
Anyone? I think this one is cooked now.

dholmes-ora reviewed (approved):
Nothing further from me. Thanks.

tstuefe reviewed (approved):
Looks still good to me.

theRealAph commented:
/integrate
